### PR TITLE
is_increasing(), is_decreasing() signature changed

### LINKED ISSUE #9947: is_decreasing(), is_increasing() and strictly functions should have `Symbol` as argument to them
`is_decreasing()`, `is_increasing()`, `is_strictly_increasing()` and `is_strictly_decreasing()` should have `x` or Symbol for as their parameter.
Because of this it will be able to deal expressions like
```
>>> is_decreasing(n + x, n, Interval(1, oo))    # here x is taken as Symbol
True
```
Also we have
```
>>> is_decreasing(1, Interval(1, 2))
AttributeError                            Traceback (most recent call last)
<ipython-input-37-9dfd81c0bb7d> in <module>()
----> 1 is_decreasing(1, Interval(1, oo))

/home/gxyd/Public/sympy/sympy/calculus/singularities.py in is_decreasing(f, interval)
    120 
    121     """
--> 122     if len(f.free_symbols) > 1:
    123         raise NotImplementedError('is_decreasing has not yet been implemented '
    124                                   'for multivariate expressions')

AttributeError: 'int' object has no attribute 'free_symbols'
```

### PR DESCRIPTION
**Pull request by gxyd**
fixes #9947

### CODE REVIEW

**Comment by leosartaj** [comment]
This doesn't work ``is_increasing(1, x)``. Similarly for the others. This can be fixed by adding a ``sympify`` call.

**Comment by leosartaj** [comment]
The ``symbol`` argument can be made a ``kwarg``. So that it is not necessary to write symbol for univariate functions.

This should work

```
>>> is_increasing(x)
True
```

**Comment by leosartaj** [comment]
Some examples for multivariate functions can be added in the docstrings.

File: sympy/calculus/singularities.py [hunk 1]
```diff
@@ -136,29 +171,40 @@ def is_strictly_decreasing(f, interval=S.Reals):
     Examples
     ========
 
-    >>> from sympy import is_decreasing
-    >>> from sympy.abc import x
+    >>> from sympy import is_strictly_decreasing
+    >>> from sympy.abc import x, y
     >>> from sympy import S, Interval, oo
-    >>> is_decreasing(1/(x**2 - 3*x), Interval.open(1.5, 3))
+    >>> is_strictly_decreasing(1/(x**2 - 3*x), Interval.open(1.5, 3))
```
Comment (gxyd) [inline sympy/calculus/singularities.py]: i have changed `is_decreasing` -> `is_strictly_decreasing`  here and in tests as well. These were probably left by mistake. @aktech can you verify. I think created these functions.

**Comment by gxyd** [comment]
@leosartaj thank you very much for all the points you made. I was though waiting for @aktech to get on to verify. But now you have seen that's really nice.

File: sympy/calculus/singularities.py [hunk 2]
```diff
@@ -57,18 +58,30 @@ def is_increasing(f, interval=S.Reals):
...
+        if len(free_sym) > 1:
+            raise NotImplementedError('is_increasing has not yet been implemented '
+                                    'for multivariate expressions')
+        if len(free_sym) == 1:
+            symbol = free_sym.pop()
+        else:
+            return True
+
+    if sym:
+        symbol = sym.popitem()[1]
```
Comment (gxyd) [inline sympy/calculus/singularities.py]: Do you think this is the best way to go?

File: sympy/calculus/singularities.py (hunk 1 above)
Comment (aktech) [inline sympy/calculus/singularities.py]: Yes, that was a mistake. Thanks for catching this.

File: sympy/calculus/singularities.py (hunk 2 above)
Comment (aktech) [inline sympy/calculus/singularities.py]: The following should work fine,

```diff
 def singularities(expr, sym):
@@ -38,7 +38,7 @@ def singularities(expr, sym):
 ########################################################################
 
 
-def is_increasing(f, interval=S.Reals):
+def is_increasing(f, interval=S.Reals, symbol=None):
     """
     Returns if a function is increasing or not, in the given
     ``Interval``.
@@ -59,10 +59,15 @@ def is_increasing(f, interval=S.Reals):
     False
 
     """
-    if len(f.free_symbols) > 1:
-        raise NotImplementedError('is_increasing has not yet been implem
-                                  'for multivariate expressions')
-    symbol = f.free_symbols.pop()
+    f = sympify(f)
+    if symbol is None:
+        symbol = [syms for syms in f.free_symbols]
+        if len(symbol) > 1:
+            raise NotImplementedError('is_increasing supports only univa
+        if not symbol:
+            return False
+        symbol = symbol.pop()
+
     df = f.diff(symbol)
```

Comment (gxyd) [inline sympy/calculus/singularities.py]: Thanks. Though i didn't saw you code before, seeing `symbol=None` thing. I saw this after pushing. 
> if not symbol:
+            return False

For this line in your code this should ` return True` for `is_increasing` and `return False` for `is_strictly_increasing` and similar for `decreasing` one's.

Comment (aktech) [inline sympy/calculus/singularities.py]: > For this line in your code this should return True for is_increasing and return False for is_strictly_increasing and similar for decreasing one's.

+1

**Comment by gxyd** [comment]
@leosartaj @aktech i think this is ready for review.

File: sympy/calculus/singularities.py [hunk 3]
```diff
@@ -117,18 +136,27 @@ def is_decreasing(f, interval=S.Reals):
...
-    if len(f.free_symbols) > 1:
-        raise NotImplementedError('is_decreasing has not yet been implemented '
-                                  'for multivariate expressions')
-    symbol = f.free_symbols.pop()
+    f = sympify(f)
+    free_sym = f.free_symbols
+
+    if symbol is None:
+        if len(free_sym) > 1:
+            raise NotImplementedError('is_increasing has not yet been implemented '
```
Comment (leosartaj) [inline sympy/calculus/singularities.py]: ``is_increasing`` should be changed to ``is_decreasing``.

File: sympy/calculus/singularities.py [hunk 4]
```diff
@@ -57,18 +58,27 @@ def is_increasing(f, interval=S.Reals):
...
-        raise NotImplementedError('is_increasing has not yet been implemented '
-                                  'for multivariate expressions')
-    symbol = f.free_symbols.pop()
+    f = sympify(f)
+    free_sym = f.free_symbols
+
+    if symbol is None:
+        if len(free_sym) > 1:
+            raise NotImplementedError('is_increasing has not yet been implemented '
+                                    'for all multivariate expressions')
```
Comment (leosartaj) [inline sympy/calculus/singularities.py]: under-indented. Maybe you can indent this more.

File: sympy/calculus/singularities.py [hunk 5]
```diff
@@ -87,18 +97,27 @@ def is_strictly_increasing(f, interval=S.Reals):
...
-        raise NotImplementedError('is_strictly_increasing has not yet been '
-                                  'implemented for multivariate expressions')
-    symbol = f.free_symbols.pop()
+    f = sympify(f)
+    free_sym = f.free_symbols
+
+    if symbol is None:
+        if len(free_sym) > 1:
+            raise NotImplementedError('is_increasing has not yet been implemented '
+                                    'for all multivariate expressions')
```
Comment (leosartaj) [inline sympy/calculus/singularities.py]: under-indented

File: sympy/calculus/singularities.py [hunk 6]
```diff
@@ -87,18 +97,27 @@ def is_strictly_increasing(f, interval=S.Reals):
...
-    if len(f.free_symbols) > 1:
-        raise NotImplementedError('is_strictly_increasing has not yet been '
-                                  'implemented for multivariate expressions')
-    symbol = f.free_symbols.pop()
+    f = sympify(f)
+    free_sym = f.free_symbols
+
+    if symbol is None:
+        if len(free_sym) > 1:
+            raise NotImplementedError('is_increasing has not yet been implemented '
```
Comment (leosartaj) [inline sympy/calculus/singularities.py]: is_strictly_increasing

File: sympy/calculus/singularities.py [hunk 7]
```diff
@@ -117,18 +136,27 @@ def is_decreasing(f, interval=S.Reals):
...
-        raise NotImplementedError('is_decreasing has not yet been implemented '
-                                  'for multivariate expressions')
-    symbol = f.free_symbols.pop()
+    f = sympify(f)
+    free_sym = f.free_symbols
+
+    if symbol is None:
+        if len(free_sym) > 1:
+            raise NotImplementedError('is_increasing has not yet been implemented '
+                                    'for all multivariate expressions')
```
Comment (leosartaj) [inline sympy/calculus/singularities.py]: under-indented

File: sympy/calculus/singularities.py [hunk 8]
```diff
@@ -136,29 +164,38 @@ def is_strictly_decreasing(f, interval=S.Reals):
...
-    if len(f.free_symbols) > 1:
-        raise NotImplementedError('is_strictly_decreasing has not yet been '
-                                  'implemented for multivariate expressions')
-    symbol = f.free_symbols.pop()
+    f = sympify(f)
+    free_sym = f.free_symbols
+
+    if symbol is None:
+        if len(free_sym) > 1:
+            raise NotImplementedError('is_increasing has not yet been implemented '
```
Comment (leosartaj) [inline sympy/calculus/singularities.py]: is_strictly_decreasing

File: sympy/calculus/singularities.py [hunk 9]
```diff
@@ -136,29 +164,38 @@ def is_strictly_decreasing(f, interval=S.Reals):
...
+                                    'for all multivariate expressions')
```
Comment (leosartaj) [inline sympy/calculus/singularities.py]: under-indented

File: sympy/calculus/singularities.py [hunk 10]
```diff
@@ -136,29 +164,38 @@ def is_strictly_decreasing(f, interval=S.Reals):
...
 
     """
-    if len(f.free_symbols) > 1:
-        raise NotImplementedError('is_strictly_decreasing has not yet been '
-                                  'implemented for multivariate expressions')
-    symbol = f.free_symbols.pop()
+    f = sympify(f)
+    free_sym = f.free_symbols
+
+    if symbol is None:
```
Comment (leosartaj) [inline sympy/calculus/singularities.py]: I see this is repeated everywhere with somewhat same error message and same structure. You should abstract this out in a private function. To be clear logic from L185 to L191 can be made a separate function.

Comment (leosartaj) [inline sympy/calculus/singularities.py]: On second thought. Leave this as it is.

File: sympy/calculus/singularities.py [hunk 11]
```diff
@@ -180,10 +217,7 @@ def is_monotonic(f, interval=S.Reals):
 
     """
     from sympy.core.logic import fuzzy_or
-    if len(f.free_symbols) > 1:
-        raise NotImplementedError('is_monotonic has not yet been '
-                                  'implemented for multivariate expressions')
```
Comment (leosartaj) [inline sympy/calculus/singularities.py]: These lines should be kept. It is a more informative error message.

File: sympy/calculus/tests/test_singularities.py [hunk 12]
```diff
@@ -8,6 +8,7 @@
 from sympy.utilities.pytest import XFAIL
 
 x = Symbol('x')
+y = Symbol('y')
```
Comment (leosartaj) [inline sympy/calculus/tests/test_singularities.py]: maybe ``x, y = symbols(' x y')``?

File: sympy/calculus/tests/test_singularities.py [hunk 13]
```diff
@@ -44,13 +50,16 @@ def test_is_decreasing():
...
-    assert is_decreasing(1/(x**2 - 3*x), Interval.Ropen(-oo, S(3)/2)) is False
-    assert is_decreasing(-x**2, Interval(-oo, 0)) is False
+    assert is_strictly_decreasing(1/(x**2 - 3*x), Interval.open(1.5, 3))
+    assert is_strictly_decreasing(1/(x**2 - 3*x), Interval.Lopen(3, oo))
+    assert is_strictly_decreasing(1/(x**2 - 3*x), Interval.Ropen(-oo, S(3)/2)) is False
+    assert is_strictly_decreasing(-x**2, Interval(-oo, 0)) is False
+    assert is_strictly_decreasing(1) is False
 
 
 def test_is_monotonic():
```
Comment (leosartaj) [inline sympy/calculus/tests/test_singularities.py]: Add a test or two showing the change.

File: sympy/calculus/tests/test_singularities.py [hunk 14]
```diff
@@ -30,6 +31,10 @@ def test_is_increasing():
     assert is_increasing(-x**2, Interval(-oo, 0))
     assert is_increasing(-x**2, Interval(0, oo)) is False
     assert is_increasing(4*x**3 - 6*x**2 - 72*x + 30, Interval(-2, 3)) is False
+    assert is_increasing(x**2 + y, Interval(1, oo), x) is True
+    t = Symbol('t', negative=True)
```
Comment (leosartaj) [inline sympy/calculus/tests/test_singularities.py]: It would be great, if you could create all the symbols at a single place.

**Comment by leosartaj** [comment]
Please do not update the PR, when someone is reviewing. It gets confusing. :smile:

**Comment by gxyd** [comment]
I have updated the error message. And addressed other comments.

**Comment by leosartaj** [comment]
You can use tools like flake8 to catch issues with indentation and such.

**Comment by leosartaj** [comment]
This looks good, +1 to merge if tests pass.

**Comment by gxyd** [comment]
Thanks, i did not know about this. Flake works good.

**Comment by leosartaj** [comment]
Looks good. Merging.

**Comment by leosartaj** [comment]
Thanks for your work.. @gxyd

**Comment by gxyd** [comment]
Thanks @leosartaj.

**Pull request merged**